core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java

```diff
@@ -81,13 +81,14 @@ public Schema record(Schema record, List<String> names, Iterable<Schema.Field> s
     List<Types.NestedField> expectedFields = struct.fields();
     for (int i = 0; i < expectedFields.size(); i += 1) {
       Types.NestedField field = expectedFields.get(i);
+      String sanitizedFieldName = AvroSchemaUtil.sanitize(field.name());
 
       // detect reordering
-      if (i < fields.size() && !field.name().equals(fields.get(i).name())) {
+      if (i < fields.size() && !sanitizedFieldName.equals(fields.get(i).name())) {
```

Comment:
Why is sanitization necessary if this fix works by removing columns from the partition columns that get attached via `JoinedRow`?

Reply:
This is happening before getting to the materialized data. Matching the columns at read time did totally lack the sanitization logic that is used at write time.